### ISSUE TITLE
Monte Carlo code spending half of its time calling the clock

### ISSUE THREAD

**Issue by amoutenet** [TRIQS/triqs#604]
Hi all,

I am currently writing a Monte Carlo code having very quick updates, and when profiling it I noticed that the code is spending 60% (!) of the time calling the now() function of the clock. I am using gcc7 and working on the baf01f commit of the unstable branch.

Looking more carefully at this problem, we noticed with Michel that this is coming from the three occurences of now() in the triqs/utility/timer.hpp header that I copy-paste below for clarity (lines of the occurences are followed by //comment, see below):
```c++
#pragma once

#include <chrono>

namespace triqs {
  namespace utility {

    class timer {
      public:
      using clock_t = std::chrono::high_resolution_clock;

      private:
      clock_t::time_point start_time;
      clock_t::duration total_time = clock_t::duration(0);
      bool running                 = false;

      public:
      void start() {
        running    = true;
        start_time = clock_t::now(); //comment
      }
      void stop() {
        total_time += clock_t::now() - start_time; //comment
        running = false;
      }
      bool is_running() const { return running; }
      operator double() const {
        std::chrono::duration<double> total_time_seconds(total_time);
        if (is_running()) total_time_seconds += clock_t::now() - start_time; //comment
        return total_time_seconds.count();
      }
    };
  } // namespace utility
} // namespace triqs
```
I tried to replace std::chrono::high_resolution_clock by std::chrono::system_clock and std::chrono::steady_clock but it didn't change anything. The only workaround we found is to comment the three lines involving now(). Fortunately, the callbacks utility is not affected by this procedure and I can still give a precise run time to my Monte Carlo.

I understand this is very deep in the standard c++ library and there is no easy solution to this problem, but I think it is good to be aware of it. 
What I would suggest is the following:
- put timing of measures optional (in my case for instance it is just the MC weight, so virtually 0)
- put the percentage optional, maybe just keeping three of them, one every 25% for instance (because it is very useful to roughly now how long a calculation will take, especially when tuning parameters before going into production mode)
- reduce as much as possible the other occurences of these calls

Cheers,
Alice

**Comment by parcollet** [comment]
Thank you Alice for analysing and reporting this.

Indeed, the clock is a system call, it will be slow whatever you do.
The timer call should not be in the inner QMC loop.

@HugoStrand @Wentzell : can one of you take care of this ?

**Comment by HugoStrand** [comment]
Dear Alice,

Thank you for reporting this. I have looked at the code and I am wondering which timers that are the cause for the 60% overhead. There is an over all timer in `mc_generic.hpp` that is timing the `run` method. However, it is only polled after each batch of `length_cycle` number of updates.

How large is your `length_cycle` in this case? Does increasing this remove the 60% time spent in the `now` function?

I thought that polling a timer after each `length_cycle` number of moves, would be a safe and moderate approach.

I am suspecting though, that this timer is not the culprit here. Since there are also timers running for each measurement, see line 64 in `mc_measure_set.hpp`.

Could you test if removing the `Timer.start()` and `Timer.stop()` calls in the `accumulate_` method of the `triqs::mc_tools::measure` class fixes the 60% time spent in calling the `now` function?

Best regards,
Hugo

**Comment by parcollet** [comment]
Assigned to Hugo. 
It is your change that introduced the pb...
Can you see directly with Alice to have the example ?
I suggest to put this feature optional ...

**Comment by HugoStrand** [comment]
I responded to this on Jul 10 with suggestions on how to diagnose the issue. Feedback from Alice is pending. Since this is a beloved feature of all the `ctqmc` codes (it is the reason why we see what measures are slow/fast) I suggest to keep this a default feature that can optionally be disabled.  Best, H

**Comment by HugoStrand** [comment]
Suggestion from the triqs meeting:
Add a boolean for turning this feature off.

**Comment by HugoStrand** [comment]
Dear @amoutenet, could you please check if the pull request #650 solves your issue? Best, H

**Comment by amoutenet** [comment]
Hi all,

I actually have a problem reproducing the issue... Before Christmas I thought it was because I was working on my Mac, but I can't reproduce it on my Linux workstation here at CCQ. Tried to remove the -march=native flag (didn't have it in the first place) but the profile still looks good.

I'm going back to my workstation in College on the given commit of the unstable branch. Maybe it's machine specific, or linked to a given optimization flag other than the -march=native (that'd be one good example of why we should always put them!).

Sorry for the inconvenience, I don't know why everything is working right now!

**Comment by Wentzell** [comment]
This problem has been addressed in 61d221ffd21fdc8f0454a6cd4b8d751d0e795fec
The timing of measurements can now be disabled.
